### ISSUE TITLE
Bring structure functionality into the core

### ISSUE THREAD

**Issue by jaeandersson** [casadi/casadi#1388]
The named structures in "casadi.tools" have proven popular and useful especially for implementing direct collocation and direct multiple shooting. I see some major issues with them however:
 * They are Python specific
 * It's not clear to me how the subexpressions are sorted, which can have a major influence on the sparsity structure of the NLP. I think the sorting should be controlled by the user.
 * I think the syntax could be more user-friendly

With #1379 now implemented, it is much less complicated than earlier to construct the NLP variable vector. Have a look at: https://github.com/casadi/casadi/blob/develop/docs/tutorials/python/src/modelica/modelica_ms.py#L163
As can be seen in that example, it is no longer necessary to preallocate an NLP vector of the right length.

A problem that remains is a convenient way to access parts of variable vector. Something along the lines of:
```python
from casadi import *
w = vertcat([MX.sym("x" + str(k), 2) for k in range(10)])
# Create "structure" with named entries from w
w0 = w(0)
w0['x', 0] = 4
w0['x', -1] = 2
```

**Comment by jaeandersson** [comment]
Also possible is something along the lines of:
```python
from casadi import *
w = vertcat([MX.sym("x" + str(k), 2) for k in range(10)])
# Create "structure" with named entries from w
w0 = DMatrix.zeros(w.sparsity())
w0[w,  0] = 4
w0[w, -1] = 2
```
That is, no new data structures are introduced, only a new type of indexing.

**Comment by jgillis** [comment]
Interesting approach, that last one.

**Comment by jaeandersson** [comment]
Actually, I realized that you can do quite a lot with existing data structures. Cf. c4d15eea81d1ba61dc459ca9890351aff0b20033

Maybe this feature is not needed at all.

**Comment by jaeandersson** [comment]
@jgillis - what do you think of just having the user make his own dictionary as in the above commit? That should be pretty transparent and flexible.

**Comment by jgillis** [comment]
You are welcome to live without it, but I refuse to resort to `range(off, off+2)`

**Comment by jaeandersson** [comment]
I'm sure we can make a short-hand for the `range(off, off+2)`.

**Comment by jgillis** [comment]
Also, struct works hierarchically

**Comment by jaeandersson** [comment]
OK, see https://github.com/casadi/casadi/blob/3e34148b43afe3f2427c7a84f52b5e6039942498/docs/tutorials/python/src/modelica/modelica_ms.py#L168 for latest iteration.
> Also, struct works hierarchically
What do you mean?

**Comment by jaeandersson** [comment]
@ghorn - what say you?

**Comment by jaeandersson** [comment]
Note that I'm not proposing to get rid of structures. Just trying to find a construction that works in C++/MATLAB, is flexible and can be understood with a minimal number of concepts.

**Comment by ghorn** [comment]
This is good functionality, but making it language specific might give more benefit than doing it all in C++. For example, in haskell the user just writes structs and dynobud magic lets you use them as design variables. But they're just haskell structs.

A C++ equivilant would be something like the following (please excuse the horrible C++ psuedo code)

```c++
// DesignVariable class would be casadi library code
template <class T>
class DesignVariable<T> {
   virtual MX toDesignVars();
   virtual static T fromDesignVars(MX & x);
   virtual static T symbolicStruct();
   virtual static MX symbolicVector();
}

// this is user code for a class representing 3 scalars
class Xyz : public DesignVariable<Xyz> {
  public:
    MX x;
    MX y;
    MX z;
}

// this is user code for a compound class
class MultiXyz : public DesignVariable<MultiXyz>{
   public:
      Xyz firstOne;
      Xyz secondOne;
}
```

So the user can write C++ code with custom structs and can magically convert to and from vectors. In haskell I take this further and provide type-indexed vectors, and parameterized types, etc. I have no idea how this would be done in C++ or python, but the point is that iwouldn't use stringly-typed casadi structs from haskell because there is already a more simple, type-safe struct implementation

**Comment by jaeandersson** [comment]
@ghorn - I'm not sure I follow.

I'm thinking that for a more advanced project, you probably want to create a custom data structure for variables and constraints anyway. But maybe that's not what you meant.

**Comment by ghorn** [comment]
What I want is for the user to create a custom data structure for variables and constraints, and then casadi can automatically convert to and from an MX. This has been invaluable in dynobud

**Comment by ghorn** [comment]
but this custom data structure should be a custom class, not some casadi class which uses strings for fields

**Comment by jaeandersson** [comment]
Right. Would you say the concept of "structures" (as in casadi.tools) is in some sense duplicating the concept of classes in Python? And in C++/Haskell it would make even less sense since these are strongly typed languages?

**Comment by jaeandersson** [comment]
As far as I'm concerned, we need a simple way that can be readily used and understood in a simple script. It should also be straightforward to extend this simple implementation using classes etc. Also, I do think that the syntax, if possible, should be as similar as possible for C++/Python/MATLAB. The user should be able to prototype algorithms in e.g. MATLAB/Python, and then be able to migrate to C++.

**Comment by ghorn** [comment]
I would say that casadi.tools structs are sort of duplicating python classes, but I don't feel that it's good or bad.

The benefit of my dynobud structs in haskell is that you can treat design variables as haskell data, and vice-versa, and you get compile errors if you index incorrectly. In Joris's python implementation you treat design variables as python dictionaries and vice-versa, and you get runtime errors if you index incorrectly, which is fine.

The difference is that in dynobud you declare the struct as haskell data, but in python you are creating this struct object at runtime. Since python is dynamically typed, i don't think it much matters how you accomplish it as long as it's simple and intuitive. If you want to add this functionality in C++, i would encourage you to make it as much like declaring a template class as possible

**Comment by ghorn** [comment]
> As far as I'm concerned, we need a simple way that can be readily used and understood in a simple script. It should also be straightforward to extend this simple implementation using classes etc. Also, I do think that the syntax, if possible, should be as similar as possible for C++/Python/MATLAB. The user should be able to prototype algorithms in e.g. MATLAB/Python, and then be able to migrate to C++.

+1, it should look like native code in whatever environment. Perhaps for matlab the cvx/cvxpy style should be copied. Users just create matrices but they can fill matlab structs with them:
```
x = MXSym(n+1,nx)
u = MXSym(n,nu)
dvs.x = x
dvs.u = u
```
This would be clean modeling but i'm not sure hwo you would order the design variables. Perhaps it wouldn't matter too much for ipopt since the linear solvers reorder internally anyways?

**Comment by ghorn** [comment]
Wait i misunderstood. I think it should look like native code, not look the same in each language. Otherwise it will be unintuitive

**Comment by jaeandersson** [comment]
I don't like the `x = MXSym(n+1,nx)` syntax. I think `x = MX.sym('x', n+1, nx)` is a perfectly valid syntax for MATLAB. The first x is the variable identifier, the second 'x' is the display-name. I think having both makes as much sense in MATLAB as in Python/C++. Even MATLAB symbolic toolbox uses the syntax `x= sym('x')`, even if there is also the shorthand `syms x`.

**Comment by ghorn** [comment]
`MX.sym` is fine with me, that wasn't my point

**Comment by jaeandersson** [comment]
I think that language-specific structures on top of CasADi is fine, but in general I think those should be part of a tool written on top of CasADi. Otherwise maintenance will be a nightmare.

**Comment by ghorn** [comment]
I can't imagine that would work for haskell, but maybe for python/matlab it's a good idea. I'll defer to others

**Comment by jgillis** [comment]
Note that I planned to brainstorm on this next week, after finishing a first optitrack iteration

**Comment by jaeandersson** [comment]
I think we should have clearly defined levels of abstraction, each having a maintainable, clearly defined API. I think "casadi.tools" blurs this a bit, even if I can accept that it's a kind of an add-on. "casadi.tools" causes a lot of pain for me though. Since it's not compiled, I don't detect when things get broken as I would do if it's C++ code. I also thinks it goes against the principle that CasADi is a "cross-platform" tool.

**Comment by jgillis** [comment]
It seems like matlab structures are not that powerful:

```
N = 10;

controls = struct('foo',12,'bar',[1 2 3]);

states = struct('pos',[1 2 3],'vel',[4 5 6],'q',3);

ocpvars = struct('X',repmat(states,N+1,1),'U',repmat(controls,N,1));

ocpvars.X(:).pos  %   get the position at every time point -- works

ocpvars.X(:).pos(1)  % get the x position at every time point -- fails
% Scalar index required for this type of multi-level indexing.

```

**Comment by ghorn** [comment]
sometimes this fails:

```
blah.foo().bar
```
but this succeeds:
```
temp = blah.foo()
ret = temp.bar
```

**Comment by jaeandersson** [comment]
Yeah, that's because of the way getitem in matlab works. For your own data structure, you can define a behavior fot e.g. `ococpvars.X(:).pos(1)`.

**Comment by jgillis** [comment]
Indeed, it seems a custom datatype is apropriate. I think ill have a shot at porting the current functionality to the core

**Comment by jgillis** [comment]
Current python struct is third generation and I am too blinded by it to see improve on the syntax. Please expose flaws in syntax

**Comment by jaeandersson** [comment]
SWIG-Matlab currently doesn't support multilevel indexing. I think this type of structure should be written in MATLAB code. Or rather, I think we should look at the implementation of casadi.tools and see what can be moved to c++, then translate the python stuff to MATLAB.

**Comment by jgillis** [comment]
Single indexing suffices for the python struct style

**Comment by jgillis** [comment]
While I like the notion of Greg's any datatype in -- any datatype out in haskell,
I think C++ is not powerful enough as a language to support that.

If you'd code up such behavior in Python,
and make an example of a class that is nested dict- and list-like, add ability to order,
and add powerful indexing facilities, then you essentially have what is `structures`.

I would think such class would be copy-pasted around like a gold standard, and never be touched.
So I wonder if it's worth the effort to support any object, while it will probably be a single one in the end.

**Comment by jgillis** [comment]
My timeline for this: should have a satisfactory implementastion by the end of the year.

**Comment by jaeandersson** [comment]
@jgillis I assume postpone?

**Comment by jgillis** [comment]
Definite postpone

**Comment by jaeandersson** [comment]
Ping @mzanon

**Comment by mzanon** [comment]
Not sure how I can contribute here, I can only report how I am using Joris style and point out that this is a very fundamental feature of casadi for me!
I report some code extract that helps illustrating why it's useful for me:
```python
states = struct_symSX( [ 'p', 'v' ])

state_unc = struct_symSX([ entry('X', shapestruct=(states,states)), ])

# This reuses previously defined variables and orders my optimisation variables in a good way
Vars  = struct_symMX([
    (
     entry("x", struct=states,    repeat=[M,N+1]),
     entry("u", struct=controls,  repeat=[M,N]),
     entry("X", struct=state_unc, repeat=[M,N+1]),# type='symm' if I want to impose that the matrix is symmetric
    ),
])

# Upper and lower bounds
lbx  = Vars( -np.inf )
ubx  = Vars(  np.inf )

# Simple way to group constraints
g = struct_MX( [
		entry("dyn",expr=dynconstr),
		entry("dynX",expr=dynconstrX),
		entry("v",expr=velconstr),
		])

# Upper and lower bounds on the generic constraints
lbg = g(0)
ubg = g(0)
lbg['v'] = -np.inf

sol = solver(arg)
Vnum = Vars(sol['x'])
gnum = g(nlp([Vnum,Pnum])[1]) # like this I can debug my code very easily!!
print gnum['dyn']

# A trick to store my simulation results in a tidy and easy to use way (could also be replaced by a dictionary)
Sim_s = struct_symMX([
    (
     entry("x",  struct=states,    repeat=[M,nSim+1]),
     entry("u",  struct=controls,  repeat=[M,nSim]),
     entry("X",  struct=state_unc, repeat=[M,nSim+1]),
    ),
])
Sim_data = Sim_s()
```

One thing which is not supported in Joris style (or which I could not figure out): I would like to be able to call
```python
Vnum['X',i,k]
```
instead of
```python
Vnum['X',i,k,'X']
```
but generating it in a similar way to what I do now.

**Comment by JonasKoenemann** [comment]
Hi!

I am using a variables class in my Matlab project to store variables in a hierarchical order. I guess its similar to the struct_symSX in python but it is independent of casadi. You can assign casadi variables to it though. You can access variables by their ID which is a string, internally i am using Hashmaps to lookup the variables for that string.
You can have a look at the class here: [Var.m](https://github.com/JonasKoenemann/optimal-control/blob/9a293f8c623c8e41bd22460686cbdaabc1bda8de/Core/Var.m)

Here some lines form the test case, how to use it:
```
state = Var('x');
state.add('p',[3,1]);
state.add('R',[3,3]);
state.add('v',[3,1]);
state.add('w',[3,1]);
controls = Var('u');
controls.add('elev',[1,1]);
controls.add('ail',[1,1]);
ocpVar = Var('ocpvar');
ocpVar.addRepeated({state,controls},5);
ocpVar.add(state);

ocpVar.set(casadi.SX.sym('v',ocpVar.size));
ocpVar.get('x',4:6).get('p').value;
```

This class could be implemented in c++, and then you provide language specific shortcuts to it, e.g. using subsref in Matlab. The last line in Matlab could be 
`ocpVar.x(4:6).p`

I am currently reimplementing the Var class to make it better designed and efficient, but it will be probably less general and more tight to the toolbox in the future..

I am not really sure if you want to implement something like that, but maybe it could be an idea!?
Jonas

**Comment by jgillis** [comment]
Awesome input, Jonas! Thanks
